game/world/triggers/movetrigger.cpp

```diff
@@ -116,177 +108,236 @@ void MoveTrigger::moveEvent() {
+  else if(state==Close) {
+    targetFrame = 0;
+    emitSound(sfxCloseStart);
+    if(prev==Open) {
+      frame = prevFrame(frame);
+      time  = keyframes[frame].ticks - time;
+      return;
+      }
+    }
   invalidateView();
-  emitSound(snd);
+  enableTicks();
   }
 
-void MoveTrigger::tick(uint64_t /*dt*/) {
-  if(state==Idle || keyframes.size()==0)
+void MoveTrigger::postProcessTrigger() {
+  if(!target.empty()) {
+    TriggerEvent e(target,vobName,TriggerEvent::T_Trigger);
+    world.triggerEvent(e);
+    }
+  if(state==Open)
+    emitSound(sfxOpenEnd);
+  if(state==Close)
+    emitSound(sfxCloseEnd);
+  if(behavior==zenkit::MoverBehavior::OPEN_TIME && state==Open) {
+    state = OpenTimed;
     return;
+    }
+  state = Idle;
+  invalidateView();
+  disableTicks();
+  }
 
-  uint32_t keySz = uint32_t(moverKeyFrames.size());
-  uint32_t maxFr = uint32_t(moverKeyFrames.size()-1);
+uint32_t MoveTrigger::nextFrame(uint32_t i) const {
+  uint32_t size = uint32_t(keyframes.size());
+  return (i+1)%size;
+  }
 
-  uint64_t maxTicks = 0;
-  for(size_t i=0; ; ++i) {
-    if(i>=keyframes.size())
-      break;
-    if(i+1>=keyframes.size() && state!=Loop && state!=NextKey)
-      break;
-    maxTicks += keyframes[i].ticks;
+uint32_t MoveTrigger::prevFrame(uint32_t i) const {
+  uint32_t size = uint32_t(keyframes.size());
+  return (i+size-1)%size;
+  }
+
+void MoveTrigger::tick(uint64_t dt) {
+  assert(state!=Idle);
+  time += dt;
+  if(state==OpenTimed) {
+    if(time<=stayOpenTime)
+      return;
+    state       = Close;
+    targetFrame = 0;
+    time       -= stayOpenTime;
     }
+  advanceAnim();
+  if(isTargetFrame())
+    postProcessTrigger(); else
+    emitSound(sfxMoving);
+  }
 
-  uint64_t dt = world.tickCount()-sAnim, ticks = 0;
-  float    a  = 0;
-  uint32_t f0 = 0, f1 = 0;
+ void MoveTrigger::advanceAnim() {
+  uint32_t f1  = (state==Close) ? prevFrame(frame) : frame;
+  uint32_t f2  = nextFrame(f1);
+  float    a   = calculateProgress(f1,f2);
+  auto     m   = calculateTransform(f1,f2,a);
+  auto     mat = pos0;
+  mat.mul(m);
+  setLocalTransform(mat);
+  }
 
-  bool finished = false;
-  switch(state) {
-    case Idle:
-      break;
-    case OpenTimed: {
-      if(sAnim+uint64_t(stayOpenTimeSec*1000)<world.tickCount()) {
-        state = Close;
-        sAnim = world.tickCount();
-        }
-      return;
-      }
-    case Open: {
-      dt = dt<maxTicks ? dt : maxTicks;
-      for(f0=0; f0+1<keyframes.size(); ++f0) {
-        if(dt<keyframes[f0].ticks)
-          break;
-        dt -= keyframes[f0].ticks;
-        }
-      f1       = std::min(f0+1,maxFr);
-      ticks    = keyframes[f0].ticks;
-      a        = ticks>0 ? float(dt)/float(ticks) : 1.f;
-      finished = f0==maxFr;
+ bool MoveTrigger::isTargetFrame() {
+  uint32_t f = (state==Close) ? prevFrame(frame) : frame;
+  if(time<keyframes[f].ticks)
+    return false;
+  if(state==Close)
+    frame = prevFrame(frame); else
+    frame = nextFrame(frame);
+  time = 0;
```

Comment:
Function name(`isTargetFrame`) suggest no side effect.
Seems `time = 0` can be moved into `postProcessTrigger` just fine.

Reply:
Time is reset when `frame` is changed. Renamed `time` to `frameTime` to make this clear.

`isTargetFrame` has been renamed to `updateFrame`. The check if mover is finished is now called after that function.